### PR TITLE
Support the TransformerEngine precision plugin with the Trainer

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The Trainer can now support this plugin easily.

Fixes https://github.com/Lightning-AI/lightning/issues/17172

cc @borda @carmocca @justusschock @awaelchli @sabetAI

----
:books: Documentation preview :books:: https://pytorch-lightning--18459.org.readthedocs.build/en/18459/

### CODE REVIEW

File: docs/source-pytorch/common/precision_intermediate.rst [hunk 1]
```diff
@@ -54,35 +54,14 @@ delivers all of these benefits while ensuring that no task-specific accuracy is
 ----
 
 
-********************
-FP16 Mixed Precision
```
Comment (carmocca) [inline docs/source-pytorch/common/precision_intermediate.rst:58]: We already talk about fp16 mixed in `precision_basic`. Added the missing paragraph there

Comment (awaelchli) [reply docs/source-pytorch/common/precision_intermediate.rst:58]: I don't think it is bad to keep this

**Review by awaelchli — APPROVED** [review]

File: docs/source-pytorch/common/precision_intermediate.rst [hunk 2]
```diff
@@ -139,12 +118,52 @@ See also: :doc:`../advanced/model_init`
...
+latest NVIDIA GPUs using 8-bit floating point (FP8) precision on Hopper GPUs, to provide better performance with lower
+memory utilization in both training and inference. It offers improved performance over half precision with no degradation in accuracy.
+
+Using TE requires replacing some of the layers in your model. Fabric automatically replaces the :class:`torch.nn.Linear`
+and :class:`torch.nn.LayerNorm` layers in your model with their TE alternatives, however, TE also offers
+`fused layers <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/pytorch.html>`__
+to squeeze out all the possible performance. If Fabric detects that any layer has been replaced already, automatic
+replacement is not done.
+
+This plugin is a mix of "mixed" and "true" precision. The computation is downcasted to FP8 precision on the fly, but
```
Comment (Borda) [inline docs/source-pytorch/common/precision_intermediate.rst:135]:
```suggestion
This plugin is a combination of "mixed" and "true" precision. The computation is downcasted to FP8 precision on the fly, but
```

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/common/precision_intermediate.rst (hunk 1 above)
Comment (carmocca) [reply docs/source-pytorch/common/precision_intermediate.rst:58]: Would you rather that I move it from basic to intermediate? Having it in both levels makes it feel duplicated. Shouldn't levels be disjoint?

Comment (carmocca) [reply docs/source-pytorch/common/precision_intermediate.rst:58]: I'll merge to get the plugin in, but I'm happy to change it or move it if you request it

**Pull request merged**